torch/optim/lars.py

```diff
@@ -0,0 +1,214 @@
+        if momentum < 0.0:
+            raise ValueError(f"Invalid momentum value: {momentum}")
+        if nesterov and (momentum <= 0 or dampening != 0):
+            raise ValueError("Nesterov momentum requires a momentum and zero dampening")
+
+        defaults = dict(
+            lr=lr,
+            momentum=momentum,
+            dampening=dampening,
+            weight_decay=weight_decay,
+            nesterov=nesterov,
+            trust_coefficient=trust_coefficient,
+            eps=eps,
+            maximize=maximize,
+            differentiable=differentiable,
+        )
+
+        super().__init__(params, defaults)
+
+    def __setstate__(self, state):
+        super().__setstate__(state)
+        for group in self.param_groups:
+            group.setdefault("nesterov", False)
+            group.setdefault("maximize", False)
+            group.setdefault("differentiable", False)
+
+    def _init_group(self, group, params_with_grad, grads, momentum_buffer_list):
+        for p in group["params"]:
+            if p.grad is None:
+                continue
+            params_with_grad.append(p)
+
+            grads.append(p.grad)
+
+            state = self.state[p]
+
+            if group["momentum"] > 0:
+                momentum_buffer_list.append(state["momentum_buffer"])
+
+    @_use_grad_for_differentiable
+    def step(self, closure=None):
+        """Performs a single optimization step.
+
+        Args:
+            closure (Callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+
+        for group in self.param_groups:
+            params_with_grad = []
+            grads = []
+            momentum_buffer_list = []
+
+            self._init_group(group, params_with_grad, grads, momentum_buffer_list)
+
+            lars(
+                params_with_grad,
+                grads,
+                momentum_buffer_list,
+                lr=group["lr"],
+                momentum=group["momentum"],
+                dampening=group["dampening"],
+                weight_decay=group["weight_decay"],
+                nesterov=group["nesterov"],
+                trust_coefficient=group["trust_coefficient"],
+                eps=group["eps"],
+                maximize=group["maximize"],
+            )
+
+            for p, momentum_buffer in zip(params_with_grad, momentum_buffer_list):
+                state = self.state[p]
+                state["momentum_buffer"] = momentum_buffer
+
+        return loss
+
+LARS.__doc__ = r"""Implements LARS algorithm.
+
+    For further details regarding the algorithm we refer to `Large Batch Training of Convolutional Networks`_.
+    """ + r"""
+    Args:
+        params (iterable): iterable of parameters to optimize or dicts defining
+            parameter groups
+        lr (float): learning rate
```

Comment:
Add default 1e-3 here

Reply:
Unless with LARS it is recommended to decide on an initial LR? From the paper, it looks like they specifically use bigger LRs.

Reply:
Yes, they experiment way bigger LR  in the paper. That's why I didn't set a default value, originally.

[Here](https://github.com/huggingface/pytorch-image-models/blob/main/timm/optim/lars.py), LR=1.0 is set.
No default value is set [here](https://github.com/keras-team/keras-contrib/blob/master/keras_contrib/optimizers/lars.py).

Reply:
Ah, maybe we should just up the default to 1.0, with a comment that larger LRs are used with LARS